### ISSUE TITLE
Help needed: Paint.NET plugins that use Menu / ContextMenu, how to get them working?

### ISSUE THREAD

**Issue by rickbrew** [dotnet/winforms#5368]
For Paint.NET 4.3, I've ported it to .NET 5, added native ARM64 support, and having been publishing public alphas. (latest build at https://forums.getpaint.net/topic/118482-paintnet-43-alpha-build-7879/ if you're curious)

One important issue that's come up is that there are some older _and very popular_ plugins, compiled circa .NET 2.0/3.5, that use `Menu` and `ContextMenu` and do not work at all. This is going to be major issue for getting people to upgrade to Paint.NET 4.3 and also .NET 5.

While I understand the reason for removing these controls, I'm at a roadblock in terms of how to get these plugins working. I can probably find the old code for these controls and build my own DLL or whatever, but how do I get the plugins to bind to the types in that DLL? `AppDomain.TypeResolve`, for instance, does not fire when it "knows" that the `Type` can't exist, e.g. it's not in a dynamic assembly. 

Short of disassembling the plugins, fixing them to use `MenuStrip` and `ContextMenuStrip`, recompiling, and then bundling the fixed DLLs with the app -- ignoring potential copyright/legal questions for the moment -- what options do I have for accomplishing this? I discussed this a bit in #lowlevel on Discord and nobody could come up with a solution. Building my own `System.Windows.Forms.dll` that includes the controls might work, but would bloat the download size by an extreme amount, and I'd really prefer to avoid doing that. I am distributing the app as framework-dependent so that I don't have to pay for the 50MB of bandwidth per install. There will be millions of these downloads, so this matters a lot. I am willing to entertain lower-level ideas like modifying the assembly in-memory or something (I'm hand-waving / guessing here).

See comments here on the need for getting these to work:
- https://forums.getpaint.net/topic/118422-paintnet-4217-alpha-build-7858-now-using-net-5/?do=findComment&comment=584690 , "That's enough errors to revert back to 4.2.16 until some corrections are made."
- https://blog.getpaint.net/2021/07/27/paint-net-4-3-alpha-build-7878/#comment-22543 , "Any news on the not working plugins? I – and many people I recommended paint.net to over the years – use them frequently. Losing them without an alternative would be a blow."

**Comment by AraHaan** [comment]
Why cant you do some sort of way to forward usage of those types to ``MenuStrip`` or even ``ContextMenuStrip``.

Bonus: if there was an .NET Version of Detours to detour construction and usage of those types to the 2 mentioned above that the old ones was removed for.

**Comment by rickbrew** [comment]
Those are intriguing ideas, but the real question is: how? How do I actually accomplish that?

Also, the plugin names and versions are knowable in advance, they're easy enough to gather. So I don't have to pay the performance cost of scanning every plugin DLL for these type references.

**Comment by AraHaan** [comment]
I think TypeResolve could work, but then return on that event return the Paint.NET Assembly and have the fake versions of those old types that was removed from winforms in there that actually "wrap" the ones they was replaced with.

**Comment by rickbrew** [comment]
As mentioned earlier, `TypeResolve` will not work, as per https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.typeresolve?view=net-5.0 👍 

> However, the TypeResolve event does not occur if the runtime knows it is not possible to find a type in certain assemblies. For example, this event does not occur if the type is not found in a static assembly because the runtime knows types cannot be added dynamically to static assemblies.

`AssemblyLoadContext` does not have an analogous event for types, only for assemblies.

**Comment by AraHaan** [comment]
That is sad I would have expected a way to hook into that on AssemblyLoadContexts (like when it cannot find those removed types in ``System.Windows.Forms.dll`` so that way you can force it to instead use the types defined in ``Paint.NET.exe``.

**Comment by ericstj** [comment]
Why don’t we add TypeForwards to an assembly that can be provided by the community?  Call it “System.Windows.Forms.Legacy”.  That will satisfy the binder without having the types in the framework. 

This is somewhat like how we handle compat shims for netfx today. Take System.Security.Permissions for example. We ship and mscorlib that forwards to it, but you need a package to get the types. Take that one step further.

**Comment by merriemcgaw** [comment]
I was thinking along the same lines @ericstj. I’m sure we can figure out how to get a package to them with the types. Let’s chat offline this week to brainstorm!

**Comment by rickbrew** [comment]
Any chance we could get that in for the next servicing release of .NET 5? 🙏🏼

**Comment by AraHaan** [comment]
Another issue with @ericstj's comment is that what will happen to removed members of ``Control`` where ``Control.ContextMenu`` or even ``Form.Menu``, etc was removed and old code depends on that property existing.

Perhaps there *should* be a language feature where one can use extensions to add ``Properties`` to another type (extension properties). Then @ericstj's idea might actually work completely.

**Comment by rickbrew** [comment]
Yeah I'm curious how @ericstj and @merriemcgaw 's ideas could work even with the type forwarding attributes and a legacy package -- the big thing I'm seeing in one of the most popular plugins is the use of the `Control.ContextMenu` property. Some other plugins do make use of `ContextMenu` with a private field and those would be fine.

I'm going to spend some time exploring the use of dnlib, https://github.com/0xd4d/dnlib , to rewrite the assembly to use `ContextMenuStrip`. IIRC converting source code from `ContextMenu` to `ContextMenuStrip` is very straightforward, so maybe I'll get (really) lucky and I can voodoo doll this into magically working. I have noooooooooo idea if this will be fruitful but I'm also really curious if it's possible, and would also mean I could ship a solution today without waiting for e.g. new language features in C# 12 or runtime features in .NET 8 😁 It could be monstrously complex and fragile though, but we'll see.

**Comment by ericstj** [comment]
My comment was only based on missing types, it won't work well for missing members as we don't have a way to forward those.  As folks have mentioned it requires assembly rewriting.  Even extension methods would require a recompile/rewrite of the customer code.

One thing that could be done is all the legacy types can be provided in `System.Windows.Forms.Legacy` as a default implementation of them that throws PNSE.  This would need to be included in the WindowsDesktop shared framework.  Then WinForms can add back all the members that expose those types, returning default values (might be a breaking change).  All these types could be annotated with obsolete to indicate to new apps not to use them.  An app could provide a `System.Windows.Forms.Legacy` with actual implementation of the legacy types, so long as it has a higher version than the framework it would win and be used.  Not sure how viable this is due to the potential breaking change as well as the non-zero cost of adding this stuff back, but I think it could be made to work.

**Comment by AraHaan** [comment]
Technically would not need to have it use a newer version if you have it as a project in your own codebase then basically have it output into the OutDir of the application that needs those types for it to win over the runtime’s copy that throws those exceptions?

Alternatively why not have the runtime version throw the exceptions if they do not explicitly opt into the old implementations of those types?

**Comment by ericstj** [comment]
> Technically would not need to have it use a newer version

I'm just stating "newer version" here since that's built into our developer experience.  For any assembly you can provide a newer version and it will be preferred over what's in the framework.  Even for framework-dependent apps you can have a app-local copy of a newer versions.  This is how we keep libraries like `System.Text.Json` evolving while still exposing new API to `netstandard2.0`.  It's not hard to make this work permanently for something like a legacy assembly, just give it a fixed assembly version in the framework (we do this, for example, for our desktop compat shims like windowsbase).

> Alternatively why not have the runtime version throw the exceptions if they do not explicitly opt into the old implementations of those types?

I don't imagine that the runtime will carry real implementations, since there was a reason this code was removed.  That's what will happen naturally if no other implementation is provided.  I think this could get more difficult if we have "complexity" in the members that needs to be present in existing WinForms types, things more than just properties for example.  Having a default implementation inbox could cause some of the problems that motivated the original removal.  You might have to design new API to be present in the legacy assembly so that WinForms can defer to it.  I imagine this would look sort of like extension methods.

Here's a diff of what's missing https://gist.github.com/ericstj/3b72c4cdfb455eb89ae88416d0054773

**Comment by rickbrew** [comment]
I managed to get the most popular plugin pack to work, @EdHarvey 's Effects, https://forums.getpaint.net/topic/18811-ed-harvey-effects-v-40-2012-02-13/

I used Mono.Cecil to rewrite the IL at load time to fix up a few simple things:
- `Control.ContextMenu` was being overridden to pass the object along to an inner control. However, neither the property nor `ContextMenu` was actually used, so I was able to safely prune it entirely.
- There were a few places where reflection was used to modify private fields of `VisualStyleRenderer`. These fields have been effectively renamed due to the use of auto-properties. For now I'm just using the new auto-generated field names, but I also know that the use of reflection isn't necessary here due to more recent changes to the `VisualStyleRenderer` class, so I may improve this shim later.

This is pretty much _the most popular plugin pack ever_, and should get me about 80% of what I need, and unblocks me from being able to release a final version of Paint.NET 4.3. Some other plugins are still using a library that makes use of `Menu` and `ContextMenu`, but they are more actively maintained and I will push the author to release updated versions.

So for now my emergency is over, but I still think that @ericstj and @merriemcgaw 's ideas and effort are worthwhile and have merit. I may also get reports of other plugins that need those types, so I can't yet close the ticket as resolved.

**Comment by merriemcgaw** [comment]
Thank you so much for the update @rickbrew! We're going to take some time next week to brainstorm how we can more broadly enable folks who are porting from .NET Framework and aren't so easily able to move from ContextMenu to ContextMenuStrip etc. I'm very glad your emergency is over so we can take the time we need to see how to support others in the same position!

**Comment by AraHaan** [comment]
If it is decided to add in a ``System.Windows.Forms.Legacy`` into WindowsDesktop and have those types throw PNSE, and bring back the properties removed from ``System.Windows.Forms`` back in 3.1 that became breaking. Will there be any plans to backport it to .NET 5, (possibly Core 3.1), and also apply it to .NET 6 too?

**Comment by merriemcgaw** [comment]
We're in the very early discussions of what might be possible. Unfortunately we don't have a plan yet - we're in the brainstorming phase. By and large we don't backport and release feature level work via servicing, and .NET 6 is shutting down. Once we have a plan we'll share it with everyone, but I can't predict yet when that would come. I promise to keep everyone posted 😺

**Comment by AraHaan** [comment]
> We're in the very early discussions of what might be possible. Unfortunately we don't have a plan yet - we're in the brainstorming phase. By and large we don't backport and release feature level work via servicing, and .NET 6 is shutting down. Once we have a plan we'll share it with everyone, but I can't predict yet when that would come. I promise to keep everyone posted 😺

I was just saying it would be great to put it in a servicing release for .NET 5 so Paint.NET 4.3 can use it in case there is more plugins Rick Brewster did not catch to Paint.NET that uses such things that was removed in Core 3.1, likewise it would be also nice to have it go in to .NET 6 so people who use LTS releases would be able to not worry about those missing types / members on winforms as well.

**Comment by merriemcgaw** [comment]
That's a very good point @AraHaan. We have some limitations we have to work within but ideally we'd have a solution for this LTS release at least. I'm excited to brainstorm it and see what we can come up with.

**Comment by merriemcgaw** [comment]
We've talked internally and while we recognize not bringing these types over teams, most are able to resolve the situation by migrating to the currently supported ones - it's products like @rickbrew's that have difficulties, when they don't have access to the source code. I'm so glad that @rickbrew was able to resolve his situation for sure! Given where we are in .NET 6, trying to rush something this late in the game really isn't a good idea. So we opted to keep an eye on the situation in .NET 7 timeframe and see how the asks from partners develop. If enough people report being totally blocked, we will investigate a solution. For the time being it's wait and see.

**Comment by rickbrew** [comment]
Well I was able to resolve it for 1 DLL that thankfully happens to cover most of what I need. I may also be able to use the technique for other DLLs as long as their use of these types is simple or elidable. There are some I know will not be doable; in theory yes, it's possible to transcribe `ContextMenu` uses over to `ContextMenuStrip`, but I would need a lot of compilers experience to dive into that.

For the sake of anyone who stumbles across this from a search or something, here's my code: https://gist.github.com/rickbrew/a340aa3188672c770fa4c79c1d14a174 ... I make no claim to its efficiency, nor its effectiveness for other scenarios :) I hereby place that code into the public domain. (the `Systrace` stuff can be removed, it's not active unless doing perf analysis)

**Comment by rickbrew** [comment]
I have made progress on making this work for other plugins that make use of `ContextMenu` and `MenuItem`, https://twitter.com/rickbrewPDN/status/1428187460203270148

I extracted the relevant classes from .NET 4.8's DLL via ILSpy and put them into a new DLL that I'll be shipping with my app, PaintDotNet.Plugins.Compatibility.dll. Then, when loading the plugin, I match against known assembly names that use these classes (e.g. `OptionBasedLibrary v0.7.9.dll` is one of them). I then run `ildasm` on the plugin (or shared library) DLL, use textual replacement to swap types, add the assembly ref, and then use `ilasm` to rehydrate it and load from a `MemoryStream`.

It's crazy, but it works! I'm quite shocked that `ContextMenu` could be transplanted with such relative ease. Load times are very slow (~850ms for that OBL v0.7.9 DLL on a Ryzen 5950X), so I'll need some kind of cache, but this means that folks will be able to upgrade to PDN 4.3 w/ .NET 5 and not having any missing plugins 😁

Thankfully none of the plugins I've found so far are using the `Control.ContextMenu` or `Form.Menu` properties.

**Comment by rickbrew** [comment]
I've published a nuget package for Menu/ContextMenu, https://www.nuget.org/packages/Unofficial.System.Windows.Forms.Legacy/0.1.0 , as per @ericstj 's naming suggestion. GitHub repo is at https://github.com/paintdotnet/System.Windows.Forms.Legacy

**Comment by AraHaan** [comment]
I think it needs to be moved to the dotnet org @rickbrew and then the .NET org can do what they need with it.

**Comment by CharlesOkwuagwu** [comment]
Hi,

Please I'm trying to migrate a project having `MenuItem` to the new `ToolStripMenuItem`
The `Index `property of `MenuItem `seems to have no equivalent in `ToolStripMenuItem`.

RE: **MenuItem.Index**
```
<summary>Gets or sets a value indicating the position of the menu item in its parent menu.</summary>
<returns>The zero-based index representing the position of the menu item in its parent menu.</returns>
```
Please advise on what to do in this case. Thanks!

**Comment by merriemcgaw** [comment]
Issue can be worked around by following the instructions [here](https://github.com/dotnet/winforms/issues/5368#issuecomment-902069621)

There are no current plans to add an official package for this purpose so we are closing this issue.
